azurerm_storage_account  - support for the `multichannel_enabled` property

alexwilcox9 opened the pull request:
Hi,

I'd like to make use of multichannel SMB shares so thought I'd have a go at adding this functionality to the storage account resource. 

https://github.com/hashicorp/terraform-provider-azurerm/issues/13715

`==> Checking that code complies with gofmt requirements...
==> Checking that Custom Timeouts are used...
==> Checking that acceptance test packages are used...
TF_ACC=1 go test -v ./internal/services/storage -run=TestAccStorageAccount_smbMultichannel -timeout 60m -ldflags="-X=github.com/hashicorp/terraform-provider-azurerm/version.ProviderVersion=acc"
=== RUN   TestAccStorageAccount_smbMultichannel
=== PAUSE TestAccStorageAccount_smbMultichannel
=== CONT  TestAccStorageAccount_smbMultichannel
--- PASS: TestAccStorageAccount_smbMultichannel (348.15s)
PASS
ok      github.com/hashicorp/terraform-provider-azurerm/internal/services/storage       348.644s`

katbyte reviewed (changes requested):
Thanks @alexwilcox9 - seems like we need to enable a feature on the storage account?
```
------- Stdout: -------
=== RUN   TestAccAzureRMStorageAccount_shareProperties
=== PAUSE TestAccAzureRMStorageAccount_shareProperties
=== CONT  TestAccAzureRMStorageAccount_shareProperties
    testcase.go:110: Step 1/2 error: Error running apply: exit status 1
        
        Error: updating Azure Storage Account `share_properties` "unlikely23exst2acct1rvfl": storage.FileServicesClient#SetServiceProperties: Failure responding to request: StatusCode=400 -- Original Error: autorest/azure: Service returned an error. Status=400 Code="FeatureNotSupportedOnStorageAccount" Message="Feature is not supported on this storage account.\nRequestId:b45ca9e4-501a-001b-7d40-c17878000000\nTime:2022-09-05T15:59:49.6466540Z"
        
          with azurerm_storage_account.test,
          on terraform_plugin_test.tf line 11, in resource "azurerm_storage_account" "test":
          11: resource "azurerm_storage_account" "test" {
        
--- FAIL: TestAccAzureRMStorageAccount_shareProperties (96.13s)
FAIL
```

- could you document how to do this in the docs?

alexwilcox9 commented:
Hi @katbyte 
This feature is only for premium storage accounts, I've added that to the documentation now.

What's confusing me now is why it complained for that test in particular as multichannel isn't enabled in that test. From a quick test it seems if you send a storage.Multichannel block on a Standard storage account it complains. Even if enabled is set to False
To try and address this I've changed the code to only send the block if it's explicitly enabled however this has caused a new problem where multichannel can't be disabled if it's already enabled (as we only send the block if it's enabled)

Have you got any tips on how to get around this?

katbyte reviewed (changes requested):
we'll need to handle the non premium account situation, so likely what needs to happen is the correct handling if the API return value in that situation

alexwilcox9 commented:
Hi @katbyte 
I think I've got it this time, both tests are passing for me. Could you please also test and let me know if the approach I've taken to deal with Standard accounts is appropriate?
Thanks!

katbyte reviewed (approved):
THanks @alexwilcox9! LGTM now 🍄
